egs/librispeech/ASR/lightweight_ctc/train.py

```diff
@@ -0,0 +1,1066 @@
+    """
+    params = get_params()
+    params.update(vars(args))
+    if params.full_libri is False:
+        params.valid_interval = 1600
+
+    fix_random_seed(params.seed)
+    if world_size > 1:
+        setup_dist(rank, world_size, params.master_port)
+
+    setup_logger(f"{params.exp_dir}/log/log-train")
+    logging.info("Training started")
+    logging.info(params)
+
+    if args.tensorboard and rank == 0:
+        tb_writer = SummaryWriter(log_dir=f"{params.exp_dir}/tensorboard")
+    else:
+        tb_writer = None
+
+    lexicon = Lexicon(params.lang_dir)
+    max_token_id = max(lexicon.tokens)
+    num_classes = max_token_id + 1  # +1 for the blank
+
+    device = torch.device("cpu")
+    if torch.cuda.is_available():
+        device = torch.device("cuda", rank)
+
+    if "lang_bpe" in str(params.lang_dir):
+        graph_compiler = BpeCtcTrainingGraphCompiler(
+            params.lang_dir,
+            device=device,
+            sos_token="<sos/eos>",
+            eos_token="<sos/eos>",
+        )
+    elif "lang_phone" in str(params.lang_dir):
+        graph_compiler = CtcTrainingGraphCompiler(
+            lexicon,
+            device=device,
+        )
+        # Manually add the sos/eos ID with their default values
+        # from the BPE recipe which we're adapting here.
+        graph_compiler.sos_id = 1
+        graph_compiler.eos_id = 1
+    else:
+        raise ValueError(
+            f"Unsupported type of lang dir (we expected it to have "
+            f"'lang_bpe' or 'lang_phone' in its name): {params.lang_dir}"
+        )
+
+    logging.info("About to create model")
+    model = MobileNetS4(
+        num_features=params.feature_dim,
+        num_classes=num_classes,
+        subsampling_factor=params.subsampling_factor,
+        first_out_channels=params.first_out_channels,
+        blocks=params.conv_blocks,
+        rnn_dim=params.rnn_dim,
+    )
+
+    if rank == 0:
+        num_param = sum([p.numel() for p in model.parameters()])
+        logging.info(f"Number of model parameters: {num_param}")
+        try:
+            import torchinfo
+            model_stats = torchinfo.summary(
+                model, input_size=(1, 1, 100, params.feature_dim))
+            logging.info(str(model_stats))
+        except ImportError:
+            logging.info('torchinfo not installed. Cannot print model summary.')
+
+    assert params.save_every_n >= params.average_period
+    model_avg: Optional[nn.Module] = None
+    if rank == 0:
+        # model_avg is only used with rank 0
+        model_avg = copy.deepcopy(model)
+
+    assert params.start_epoch > 0, params.start_epoch
+    checkpoints = load_checkpoint_if_available(
+        params=params, model=model, model_avg=model_avg
+    )
+
+    model.to(device)
+    if world_size > 1:
+        logging.info("Using DDP")
+        model = DDP(model, device_ids=[rank])
+
+    optimizer = Eve(model.parameters(), lr=params.initial_lr)
```

Comment:
From the documentation of `Eve`, it is designed to be used in `scaled` model. I don't know if `Eve` would benefit you model. You might want to try original Adam or AdamW, up to you, though.

Reply:
> From the documentation of `Eve`, it is designed to be used in `scaled` modules. I don't know if `Eve` would benefit your model. You might want to try original Adam or AdamW, up to you, though.

Thanks. I was wondering that too. Will try AdamW. Is Eden supposed to be used together with Eve?

Reply:
Yes, Eden has to be used with Eve, it is not suitable for Adam-style updates.
For a convolutional-style network with Adam, I'd probably aim for something that starts out constant and then maybe decays slowly.  E.g. halve the LR on epoch 10 and then on epoch 20, and then leave constant, I think that would be enough decrease.  The absolute value of the learning rate probably will have relatively little effect, I forget what normal values are.

Reply:
> Yes, Eden has to be used with Eve, it is not suitable for Adam-style updates. For a convolutional-style network with Adam, I'd probably aim for something that starts out constant and then maybe decays slowly. E.g. halve the LR on epoch 10 and then on epoch 20, and then leave constant, I think that would be enough decrease. The absolute value of the learning rate probably will have relatively little effect, I forget what normal values are.

Thank you. I've seen some icefall recipes using AdamW with initial lr=3e-3, and
scheduler = StepLR(optimizer, step_size=8, gamma=0.1) or 
scheduler = StepLR(optimizer, step_size=2, gamma=0.8)
The result is probably not very sensitive to initial LR given Adam is adaptive.